<a href="https://colab.research.google.com/github/shilongjaycui/mathematical-programming/blob/main/modular_arithmetic.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Problem 1

In [ ]:
def num_points(a,b,p):
  s = 0
  for x in range(p): #incorrect format: for x < p
    for y in range(p): #incorrect format: for y < p
      if (y**2 - x**3 - a*x - b) % p == 0:
        s += 1
  return s

print(num_points(1,1,5))
print(num_points(0,1,97))

8
83


Problem 2

In [ ]:
def num_points2(a,b,c,d,e,f,p):
  s = 0
  for x in range(p): #incorrect format: for x < p
    for y in range(p): #incorrect format: for y < p
      if y**2 % p == (a*x**5 + b*x**4 + c*x**3 + d*x**2 + e*x + f) % p: #don't forget to do % p
        s += 1
  return s

def no_solution(): #a function can have no inputs
  for a in range(1,11):
    for b in range(1,11):
      for c in range(1,11):
        for d in range(1,11):
          for e in range(1,11):
            for f in range(1,11):
              if num_points2(a,b,c,d,e,f,11) == 0:
                return a,b,c,d,e,f
              
print(no_solution())

(1, 1, 7, 8, 3, 8)


Problem 3

In [ ]:
import sympy as sp

for n in range(3,501): #range(x,y) includes x but not y
  pseudo = 0
  if (2**(n-1) - 1) % n == 0:
    print(n)
    if sp.isprime(n) == False:
      pseudo = n
      print('The composite number is', pseudo)

3
5
7
11
13
17
19
23
29
31
37
41
43
47
53
59
61
67
71
73
79
83
89
97
101
103
107
109
113
127
131
137
139
149
151
157
163
167
173
179
181
191
193
197
199
211
223
227
229
233
239
241
251
257
263
269
271
277
281
283
293
307
311
313
317
331
337
341
The composite number is 341
347
349
353
359
367
373
379
383
389
397
401
409
419
421
431
433
439
443
449
457
461
463
467
479
487
491
499


These n are mostly prime numbers, but there's a "pseudo-prime" (341) that acts like prime numbers but is in fact composite numbers. 

Problem 4

In [ ]:
for p in range(21,100):
  if (2**(p-1) - 1) % p == 0:
    for n in range(1, p):
      k = 1
      while (n**k - 1) % p != 0:
        k += 1
      print(p,k)

23 1
23 11
23 11
23 11
23 22
23 11
23 22
23 11
23 11
23 22
23 22
23 11
23 11
23 22
23 22
23 11
23 22
23 11
23 22
23 22
23 22
23 2
29 1
29 28
29 28
29 14
29 14
29 14
29 7
29 28
29 14
29 28
29 28
29 4
29 14
29 28
29 28
29 7
29 4
29 28
29 28
29 7
29 28
29 14
29 7
29 7
29 7
29 28
29 28
29 2
31 1
31 5
31 30
31 5
31 3
31 6
31 15
31 5
31 15
31 15
31 30
31 30
31 30
31 15
31 10
31 5
31 30
31 15
31 15
31 15
31 30
31 30
31 10
31 30
31 3
31 6
31 10
31 15
31 10
31 2
37 1
37 36
37 18
37 18
37 36
37 4
37 9
37 12
37 9
37 3
37 6
37 9
37 36
37 12
37 36
37 9
37 36
37 36
37 36
37 36
37 18
37 36
37 12
37 36
37 18
37 3
37 6
37 18
37 12
37 18
37 4
37 36
37 9
37 9
37 36
37 2
41 1
41 20
41 8
41 10
41 20
41 40
41 40
41 20
41 4
41 5
41 40
41 40
41 40
41 8
41 40
41 5
41 40
41 5
41 40
41 20
41 20
41 40
41 10
41 40
41 10
41 40
41 8
41 40
41 40
41 40
41 10
41 4
41 20
41 40
41 40
41 20
41 5
41 8
41 20
41 2
43 1
43 14
43 42
43 7
43 42
43 3
43 6
43 14
43 21
43 21
43 7
43 42
43 21
43 21
43 21
43 7
43 21
43 42
43 42
43 4

k is a factor of p - 1, and all factors of p - 1 are listed as k.